tsfc/coffee_mode.py

```diff
@@ -0,0 +1,325 @@
+    connections = []
+    # add connections (list of tuples, items in each tuple form a product)
+    for monomial in monomial_sum:
+        connection = []
+        for atomic in monomial.atomics:
+            if atomic not in atomic_index:
+                atomic_index[atomic] = next(index)
+            connection.append(atomic_index[atomic])
+        connections.append(tuple(connection))
+
+    if len(atomic_index) == 0:
+        return ((), ())
+    if len(atomic_index) == 1:
+        return ((next(iterkeys(atomic_index)), ), ())
+
+    # set up the ILP
+    import pulp as ilp
+    ilp_prob = ilp.LpProblem('gem factorise', ilp.LpMinimize)
+    ilp_var = ilp.LpVariable.dicts('node', range(len(atomic_index)), 0, 1, ilp.LpBinary)
+
+    # Objective function
+    # Minimise number of factors to pull. If same number, favour factor with larger extent
+    big = 1e20  # some arbitrary big number
+    ilp_prob += ilp.lpSum(ilp_var[index] * (big - index_extent(atomic, argument_indices))
+                          for atomic, index in iteritems(atomic_index))
+
+    # constraints
+    for connection in connections:
+        ilp_prob += ilp.lpSum(ilp_var[index] for index in connection) >= 1
+
+    ilp_prob.solve()
+    if ilp_prob.status != 1:
+        raise RuntimeError("Something bad happened during ILP")
+
+    def optimal(atomic):
+        return ilp_var[atomic_index[atomic]].value() == 1
+
+    optimal_atomics = filter(optimal, iterkeys(atomic_index))
+    return tuple(optimal_atomics)
+
+
+def factorise_atomics(monomial_sum, sum_indices_frozen_set, optimal_atomics, argument_indices):
+    """Group and factorise monomials using a list of atomics as common
+    subexpressions. Create new monomials for each group and optimise them recursively.
+
+    :arg monomial_sum: a :class:`MonomialSum` object, all of its monomials should
+                       have the same sum indices
+    :arg sum_indices_frozen_set: frozen set of sum indices of the monomials in monomial_sum
+    :arg optimal_atomics: list of tuples of atomics to be used as common subexpression
+                          and the frozenset of their sum indices
+    :arg argument_indices: tuple of argument indices
+
+    :returns: a factorised :class:`MonomialSum` object, or the original object
+    if no changes are made
+    """
+    if not optimal_atomics:
+        return monomial_sum
+    if len(monomial_sum) < 2:
+        return monomial_sum
+    new_monomial_sum = MonomialSum()
+    # Group monomials with respect to each optimal atomic
+    factor_group = OrderedDict()
+    for monomial in monomial_sum:
+        for oa in optimal_atomics:
+            if oa in monomial.atomics:
+                # Add monomial to the list of corresponding optimal atomic
+                factor_group.setdefault(oa, []).append(monomial)
+                break
+        else:
+            # Add monomials that do no have argument factors to new MonomialSum
+            new_monomial_sum.add(*monomial)
```

Comment:
Ah, I think I start to understand what is going on here. It is the grouping pattern again! For at least the third time in the same pull request! Which means we must write a generic higher-order function to capture this pattern, just like a real Haskell programmer would be do.

In this concrete case, you are grouping monomials again, but this time the grouping key is the `set.intersection` with `optimal_atomics`. (Which you seem to assume to contain one or zero "atomics".)